### PR TITLE
JDK-8316708: Augment WorstCaseTests with more cases

### PR DESCRIPTION
**Pull request by jddarcy**
A new paper

 "Accuracy of Mathematical Functions in Single, Double, Double Extended, and Quadruple Precision"
by Brian Gladman, Vincenzo Innocente and Paul Zimmermann
https://members.loria.fr/PZimmermann/papers/accuracy.pdf

details the inputs with generate the worst-case observed errors in different math library implementations. The FDLIBM-related worst cases should be added to the test suite.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316708](https://bugs.openjdk.org/browse/JDK-8316708): Augment WorstCaseTests with more cases (**Enhancement** - P4)

### Reviewers
 * [Raffaello Giulietti](https://openjdk.org/census#rgiulietti) (@rgiulietti - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15879/head:pull/15879` \
`$ git checkout pull/15879`

Update a local copy of the PR: \
`$ git checkout pull/15879` \
`$ git pull https://git.openjdk.org/jdk.git pull/15879/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15879`

View PR using the GUI difftool: \
`$ git pr show -t 15879`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15879.diff">https://git.openjdk.org/jdk/pull/15879.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15879#issuecomment-1730832212)

### CODE REVIEW

**Comment by jddarcy** [comment]
Since the error bars for sinh and cosh are larger, I didn't include those cases. FDLIBM pow has a bug where the actual error is outside of the spec'ed error bounds so that case is that included in this update.

File: test/jdk/java/lang/Math/WorstCaseTests.java [hunk 1]
```diff
@@ -296,6 +320,9 @@ private static int testWorstTan() {
             {+0x1.D696BFA988DB9p-2,     +0x1.FAC71CD34EEA6p-2},
             {+0x1.46AC372243536p-1,     +0x1.7BA49F739829Ep-1},
             {+0x0.A3561B9121A9Bp+0,     +0x0.BDD24FB9CC14Fp+0},
+
+            // Worst-case observed error
+            {0x1.3f9605aaeb51bp+21,     -0x1.9678ee5d64935p-1},
```
Comment (rgiulietti) [inline test/jdk/java/lang/Math/WorstCaseTests.java:325]: The given expected value seems to contradict the introductory comment.
The exact value y meets `-0x1.9678ee5d64935p-1` < y < `-0x1.9678ee5d64934p-1`, and is closer to `-0x1.9678ee5d64935p-1`.
Thus, the rounded value of y is `-0x1.9678ee5d64935p-1`, but its truncated value is `-0x1.9678ee5d64934p-1`.
This should be the expected value, but then the test fails.

I don't think that the test logic can support errors > 1 ulp, as is the case here.
Perhaps, rather than a single expected value, there should be explicit lower and upper bounds.

File: test/jdk/java/lang/Math/WorstCaseTests.java [hunk 2]
```diff
@@ -52,6 +52,12 @@
  * JDK implementation complies with a 1 ulp bound on the worst-case
  * values.  Therefore, no addition leeway is afforded when testing
  * sinh and cosh.
+ *
+ * Additional worst-case observed error inputs for the FDLIBM-dervied
```
Comment (rgiulietti) [inline test/jdk/java/lang/Math/WorstCaseTests.java:56]:
```suggestion
 * Additional worst-case observed error inputs for the FDLIBM-derived
```

File: test/jdk/java/lang/Math/WorstCaseTests.java [hunk 3]
```diff
@@ -267,6 +288,9 @@ private static int testWorstAcos() {
             {+0x1.4182199998587p-1,     +0x1.C8A538AE83D1Fp-1},
             {+0x1.E45A1C93651ECp-1,     +0x1.520DC553F6B23p-2},
             {+0x1.F10FC61E2C78Fp-1,     +0x1.EFEEF61D39AC1p-3},
+
+            // Worst-case observed error
+            {-0x1.0068b067c6feep-1,     +0x1.0c335e2f0727p1},
```
Comment (rgiulietti) [inline test/jdk/java/lang/Math/WorstCaseTests.java:293]: Similar considerations as for asin above.
The "expected" value should be `0x1.0c335e2f0726fp1`.

File: test/jdk/java/lang/Math/WorstCaseTests.java [hunk 4]
```diff
@@ -243,6 +261,9 @@ private static int testWorstCos() {
             {+0x1.7CB7648526F99p-1,     +0x1.78DAF01036D0Cp-1},
             {+0x1.C65A170474549p-1,     +0x1.434A3645BE208p-1},
             {+0x1.6B8A6273D7C21p+0,     +0x1.337FC5B072C52p-3},
+
+            // Worst-case observed error
+            {-0x1.34e729fd08086p+21,    +0x1.6a6a0d6a17f0fp-1},
```
Comment (rgiulietti) [inline test/jdk/java/lang/Math/WorstCaseTests.java:266]: Both `Math.cos` and `StrictMath.cos` produce the correctly rounded result here.
I don't know why the paper says otherwise. Perhaps OpenLibm is not exactly fdlibm.

File: test/jdk/java/lang/Math/WorstCaseTests.java (hunk 1 above)
Comment (jddarcy) [reply test/jdk/java/lang/Math/WorstCaseTests.java:346]: For FDLIBM tan, the stated error in the Math.tan spec is 1 ulp and the FDLIBM sources say tan is "nearly rounded," which could reasonably be interpreted as meaning within 1 ulp. However, the reported error by the paper is 1.02 ulps.

As you note here, the current testing methodology can only really deal with at most a 1 ulp error, assuming the expected value is at the lower endpoint of the range.

To avoid any lurking errors in the FDLIBM port to Java, I generated the expected numbers running jshell on JDK 20, which uses the mostly C-based FDLIBM. For a number of cases I had to decrement the expected value for the test to pass against Math and StrictMath. The decremented value and its successor may or may not bracket the exact value; I didn't verify that.

In other words, there may be other bugs in one or both math libraries the test is detecting.

So far, I've only tried running the test locally, but this would need a cross-platform run being before pushed to cover all the intrinsics that may be in use on the full set of supported platforms.

**Comment by jddarcy** [comment]
Keep alive.

**Comment by rgiulietti** [comment]
ping...

File: test/jdk/java/lang/Math/WorstCaseTests.java (hunk 4 above)
Comment (jddarcy) [reply test/jdk/java/lang/Math/WorstCaseTests.java:281]: > Both `Math.cos` and `StrictMath.cos` produce the correctly rounded result here. I don't know why the paper says otherwise. Perhaps OpenLibm is not exactly fdlibm.

I've looked a bit over the OpenLibm changelog. They've added a few special cases for exp and pow as leaset. OpenLibm itself is a direct derivative of one of the BSD math libraries, which in turn was derived from FDLIBM.

**Comment by jddarcy** [comment]
Upon further reflection, I think it is worthwhile to include the worst-case tests from the more of the other math library implementations to better test potential intrinsifications of the various methods.

File: test/jdk/java/lang/Math/WorstCaseTests.java (hunk 1 above)
Comment (jddarcy) [reply test/jdk/java/lang/Math/WorstCaseTests.java:364]: > For FDLIBM tan, the stated error in the Math.tan spec is 1 ulp and the FDLIBM sources say tan is "nearly rounded," which could reasonably be interpreted as meaning within 1 ulp. However, the reported error by the paper is 1.02 ulps.
> 
> As you note here, the current testing methodology can only really deal with at most a 1 ulp error, assuming the expected value is at the lower endpoint of the range.
> 
> To avoid any lurking errors in the FDLIBM port to Java, I generated the expected numbers running jshell on JDK 20, which uses the mostly C-based FDLIBM. For a number of cases I had to decrement the expected value for the test to pass against Math and StrictMath. The decremented value and its successor may or may not bracket the exact value; I didn't verify that.
> 
> In other words, there may be other bugs in one or both math libraries the test is detecting.
> 
> So far, I've only tried running the test locally, but this would need a cross-platform run being before pushed to cover all the intrinsics that may be in use on the full set of supported platforms.

Subsequently, the allowable error bound for tan was bumped up to 1.25 ulps to cover the observed 1.02 ulp error under JDK-8326530: Widen allowable error bound of Math.tan.

File: test/jdk/java/lang/Math/WorstCaseTests.java (hunk 3 above)
Comment (jddarcy) [reply test/jdk/java/lang/Math/WorstCaseTests.java:329]: For the new worst-case conditions I've pushed, if there is a "// check" comment, the Math and StrictMath return values were the same and I did not yet check which exact floating-point value should be used as the reference point.

File: test/jdk/java/lang/Math/WorstCaseTests.java (hunk 4 above)
Comment (jddarcy) [reply test/jdk/java/lang/Math/WorstCaseTests.java:310]: > > Both `Math.cos` and `StrictMath.cos` produce the correctly rounded result here. I don't know why the paper says otherwise. Perhaps OpenLibm is not exactly fdlibm.
> 
> I've looked a bit over the OpenLibm changelog. They've added a few special cases for exp and pow as least. OpenLibm itself is a direct derivative of one of the BSD math libraries, which in turn was derived from FDLIBM.

The BSDs look to have changed/improved the kernel cos computation in their fork.

**Comment by jddarcy** [comment]
To improve the "fingerprinting" coverage of the StrictMath tests, I've added test cases where the worst-case of a non-FDLIBM library is larger than the FDLIBM worst-case. Assuming reasonable methodology of the paper, the output of the other math library must differ from FDLIBM at such a point.

**Comment by jddarcy** [comment]
> To improve the "fingerprinting" coverage of the StrictMath tests, I've added test cases where the worst-case of a non-FDLIBM library is larger than the FDLIBM worst-case. Assuming reasonable methodology of the paper, the output of the other math library must differ from FDLIBM at such a point.

In a subsequent push, I'll fill those in for the functions not already so updated (cbrt, expm1, log10, log1p, etc.)

File: test/jdk/java/lang/StrictMath/HypotTests.java [hunk 5]
```diff
@@ -705,6 +705,11 @@ static int testHypot() {
             {0x0.0000000000001p-1022, 0x0.0000000000001p-1022, 0x0.0000000000001p-1022},
             {0x0.0000000000002p-1022, 0x0.0000000000001p-1022, 0x0.0000000000002p-1022},
             {0x0.0000000000003p-1022, 0x0.0000000000002p-1022, 0x0.0000000000004p-1022},
+
+            // Empirical worst-case points in other libraries with
+            // larger worst-case errors than FDLIBM
+            {0x0.00000000039a2p-1022, 0x0.0000000000001p-1022, 0x0.00000000039a2p-1022},
```
Comment (rgiulietti) [inline test/jdk/java/lang/StrictMath/HypotTests.java:711]: I can't find this one on the paper.
```suggestion
            {-0x0.fffffffffffffp-1022, 0x0.0000000000001p-1022, 0x0.fffffffffffffp-1022},
```

File: test/jdk/java/lang/StrictMath/Log1pTests.java [hunk 6]
```diff
@@ -209,6 +209,14 @@ static int testLog1p() {
...
             {0x1.249ad2594989p332,      0x1.cc845b54b54a6p7},
+
+            // Empirical worst-case points in other libraries with
+            // larger worst-case errors than FDLIBM
+            {-0x1.2bf183e0344b2p-2,     -0x1.62ebb44459d79p-2},
+            {-0x1.2bf32aaf122e2p-2,     -0x1.62ee0a3a4baf9p-2},
+            {-0x1.8000000000000p-53,    -0x1.8000000000001p-53},
+            {-0x1.2e496d25897ecp-2,     -0x1.663d81cb08f56p-2},
+            {-0x1.ffffffbaefe27p-2,     -0x1.62e42faa93817p-1},
         };
```
Comment (rgiulietti) [inline test/jdk/java/lang/StrictMath/Log1pTests.java:220]: I think there's another case
```suggestion
            {-0x1.5efad5491a79bp-1022,     -0x1.5efad5491a79bp-1022},
        };
            
```

File: test/jdk/java/lang/StrictMath/LogTests.java [hunk 7]
```diff
@@ -68,6 +68,10 @@ static int testLog() {
             {0x1.0000000744b3ap632,  0x1.b611ab2bd53cep8},
             {0x1.000000037d81fp766,  0x1.0979b1dbc4a42p9},
             {0x1.000000024028p991,   0x1.577455642bb92p9},
+            // Empirical worst-case points
+            {0x1.0ffea3878db6bp+0,   0x1.f07a0cca521fp-5},
+            {0x1.490af72a25a81p-1,  -0x1.c4bf7ae48f078p-2},
+            {0x1.69e7aa6da2df5p-1,  -0x1.634508c9adfp-2},
```
Comment (rgiulietti) [inline test/jdk/java/lang/StrictMath/LogTests.java:74]: There are no libraries that have worse errors than OpenLibm, so I'm wondering what these values are good for?

File: test/jdk/java/lang/StrictMath/TrigTests.java [hunk 8]
```diff
@@ -164,6 +164,11 @@ private static int testAgainstTranslitSin() {
             {0x1.00000000000fcp299,  0x1.78ad2fd7aef78p-1},
             {0x1.0000000000002p300, -0x1.1adaf3550facp-1},
             {0x1.00000000001afp1023, 0x1.d1c804ef2eeccp-1},
+            // Empirical worst-case points
+            {-0x1.f8b791cafcdefp+4,  -0x1.073ca87470dfap-3},
+            {-0x1.0e16eb809a35dp+944, 0x1.b5e361ed01dadp-2},
+            {-0x1.842d8ec8f752fp+21, -0x1.6ce864edeaffep-1},
+            {-0x1.1c49ad613ff3bp+19, -0x1.fffe203cfabe1p-2},
```
Comment (rgiulietti) [inline test/jdk/java/lang/StrictMath/TrigTests.java:171]: Some of these cases are for libraries that have _better_ worst case errors than OpenLibm.
Are they needed here?

File: test/jdk/java/lang/StrictMath/HypotTests.java (hunk 5 above)
Comment (jddarcy) [reply test/jdk/java/lang/StrictMath/HypotTests.java:711]: > I can't find this one on the paper.

Good catch; must have been a cut and paste error on my part.

File: test/jdk/java/lang/StrictMath/LogTests.java (hunk 7 above)
Comment (jddarcy) [reply test/jdk/java/lang/StrictMath/LogTests.java:74]: > There are no libraries that have worse errors than OpenLibm, so I'm wondering what these values are good for?

When I was working on updating the worst-case tests for Math, I would check the input values in Math.foo() and StrictMath.foo() and if they differed pick the smaller one as the reference value. The those inputs that differ, the input is a fingerprint marker for FDLIBM and I added cases to the corresponding StrictMath test.

**Comment by jddarcy** [comment]
Using high-precision results from http://www.apfloat.org/calculator/, bounds in the worst-case test adjusted to the bracketing floating-point value closer to zero. Updated tests run successfully on Oracle cross-platform build and test system.

**Comment by rgiulietti** [comment]
Tests.java L:111 should read
```
     * @param d floating-point number whose exponent is to be extracted
```

**Review by rgiulietti — APPROVED** [review]
